sys/sysent.c

```diff
@@ -24,8 +28,69 @@ int sys_write(thread_t *td, syscall_args_t *args) {
   return -EBADF;
 }
 
+/* This is just a stub. A full implementation of this syscall will probably
+   deserve a separate file. */
+/* Note that this sbrk implementation does not actually extend .data section,
+   because we have no guarantee that there is any free space after .data in the
+   memory map. But it does not matter much, because no application would assume
+   that we are actually expanding .data, it will use the pointer returned by
+   sbrk. */
+int sys_sbrk(thread_t *td, syscall_args_t *args) {
+  size_t increment = (size_t)args->args[0];
+
+  kprintf("[syscall] sbrk(%zu)\n", increment);
+
+  assert(td->td_uspace);
+
+  if (!td->td_uspace->td_brk_entry) {
```

Comment:
I don't think there's such thing as *break segment*. Here's an excerpt from *brk(2)* manual:

> brk() and sbrk() change the location of the program break, which defines the end of the process's data segment (i.e., the program break is the first location after the end of the uninitialized data segment).

I'm not sure if there's any software that relies on the fact that *program break* should start immediately after `_end` linker symbol. There's no trace of `sbrk` in FreeBSD's [vmspace](http://bxr.su/FreeBSD/sys/vm/vm_map.h#234). OTOH in Linux [mm_struct](http://lxr.free-electrons.com/source/include/linux/mm_types.h#L396) holds `start_brk` and `brk` fields.

Short program running on Linux suggests *sbrk* doesn't have to be mapped after `.bss`:
```c
#include <unistd.h>
#include <stdio.h>

extern char _edata[0];
extern char _end[0];

int main() {
  void *_ebrk = sbrk(0);
  printf("edata = %p\n", _edata);
  printf("end = %p\n", _end);
  printf("ebrk = %p\n", _ebrk);
  pause();
  return 0;
}
```

```
# ./test
edata = 0x5595e6b77040
end = 0x5595e6b77048
ebrk = 0x5595e856f000
```

```
# pmap $(pgrep test)
18096:   ./test
00005595e6976000      4K r-x-- test
00005595e6b76000      4K r---- test
00005595e6b77000      4K rw--- test
00005595e856f000    132K rw---   [ anon ]
```